File: typescript/src/port.ts
import type { Payload } from "./payload";
import type { SendResult } from "./result";
import type { Severity } from "./severity";

/**
 * Transport-only notification port. CONTRACT.md §Port.
 * The concrete implementation is `NotificationClient`; consumers may type
 * against this interface to substitute the implementation in tests.
 */
export interface NotificationPort {
  /**
   * Multicast the payload to every configured adapter in parallel.
   * See CONTRACT.md §Methods → `send` for the full semantics.
   */
  send(severity: Severity, payload: Payload): Promise<SendResult>;
}

File: python/src/contriwork_notifications/port.py
"""Public port: NotificationPort Protocol. CONTRACT.md §Port."""

from __future__ import annotations

from typing import Protocol, runtime_checkable

from .payload import Payload
from .result import SendResult
from .severity import Severity


@runtime_checkable
class NotificationPort(Protocol):
    """Transport-only notification port.

    The concrete implementation is :class:`contriwork_notifications.NotificationClient`,
    which also satisfies this Protocol so callers can type against the
    abstraction.
    """

    async def send(self, severity: Severity, payload: Payload) -> SendResult:
        """Multicast the payload to every configured adapter in parallel.

        See CONTRACT.md §Methods → ``send`` for the full semantics.
        """
        ...
